### PR TITLE
Adds pre commit function to studiorc and gofmt to verify pipeline

### PR DESCRIPTION
**Pull request by jonsmorrow**
## Description
* Adds static_tests fucntion to .studiorc that runs gofmt.
* Adds verify pipeline step to call static_tests to ensure code matches gofmt.
* Adds pre_commit_verify to .studiorc that runs all checks that shoudl pass before submitting a pr.
* Updates unit_tests and integration_tests in .studiorc to return number of failed tests for reporting in pre_commit_verify.

Signed-off-by: Jon Morrow <jmorrow@chef.io>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [X] I have read the **CONTRIBUTING** document.
- [X] I have run the pre-merge tests locally and they pass.
- [X] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: .studiorc [hunk 1]
```diff
@@ -127,3 +176,104 @@ function install_cookstyle() {
   RUBY_BIN="$(hab pkg path core/ruby)/bin"
   export PATH="$RUBY_BIN:$PATH"
 }
+# Adding auto tab complete
+complete -F _install_cookstyle
+
+document "pre_commit_verify" <<DOC
+  Runs all verify steps on local code before commiting or creating pr.
+DOC
+function pre_commit_verify() {
```
Comment (tyler-ball) [inline .studiorc]: I don't see this function used anywhere. Where is the plan to run it?

Comment (jonsmorrow) [reply .studiorc]: In studio prior to submitting a pr.

**Comment by jonsmorrow** [comment]
render1582138003341

File: .studiorc [hunk 2]
```diff
@@ -65,6 +67,33 @@ function gocode_generation() {
...
+DOC
+function static_tests() {
+  install_if_missing core/go gofmt
+
+  ( cd /src || return 1
+    local gofiles
+    gofiles=$(find ./ -maxdepth 3 -name '*.go')
+    [ -z "$gofiles" ] && return 0
+
+    unformatted=$(gofmt -l $gofiles)
```
Comment (afiune) [inline .studiorc]: `gofmt` not only verifies format but also when the code doesn't compile. If you have an error then
you will see the following output:
```
[65][default:/src:0]# static_tests
pkg/formatter/csv_test.go:35:2: expected declaration, found '}'
pkg/formatter/csv_test.go:40:2: expected declaration, found actual
pkg/formatter/csv_test.go:47:2: expected declaration, found cbStatus
pkg/formatter/csv_test.go:188:2: expected declaration, found actual
  hab-studio: Passed!!
```

Could we also verify if there are errors from the Go files?

File: .studiorc [hunk 3]
```diff
@@ -65,6 +67,33 @@ function gocode_generation() {
...
+    if [ -z "$unformatted" ]
+    then
+      log_line "Passed!!"
```
Comment (afiune) [inline .studiorc]: nit: use color functions.[ `green [message]`](https://github.com/chef/studio-common/blob/master/dot-studio/logging#L40)
```suggestion
      log_line "$(green Passed\!)"
```

File: .studiorc [hunk 4]
```diff
@@ -127,3 +162,100 @@ function install_cookstyle() {
...
+document "pre_commit_verify" <<DOC
+  Runs all verify steps on local code before commiting or creating pr.
+DOC
+function pre_commit_verify() {
+  ( cd /src || return 1
+    log_line "Running Static Tests..."
+    static_errors=$(static_tests 2>&1 > /dev/null)
+    static_offenses=$?
+    
+    echo -en "\033[1A\033[14C\033[0KStatic Tests:\033[6C"
```
Comment (afiune) [inline .studiorc]: Lots of nice colors, I would suggest using the standards, here are some examples:

https://github.com/chef/studio-common/blob/master/dot-studio/logging#L19
# ANSI Escape Codes:

Just so you can be pretty in your messages :smile:

### Example with `echo`
```
echo -e "I $(red love) $(blue chef) $(yellow friends)!"
```

### Or using the `log_line` helper:
```
log_line "I $(red love) $(green chef) $(yellow friends)!"
```

image

File: .studiorc [hunk 5]
```diff
@@ -65,6 +67,33 @@ function gocode_generation() {
...
+    [ -z "$gofiles" ] && return 0
+
+    unformatted=$(gofmt -l $gofiles)
+    if [ -z "$unformatted" ]
+    then
+      log_line "Passed!!"
+      return 0
+    fi
+
+    echo >&2 "Failure: Go files must be formatted with gofmt. Please run:"
```
[the remark this answers is not in the document]
Comment (jonsmorrow) [reply .studiorc]: this bit was pulled from automate. I'll look at your suggestion.

File: .studiorc [hunk 6]
```diff
@@ -65,6 +67,33 @@ function gocode_generation() {
...
 
+document "static_tests" <<DOC
+  Run static code tests
+DOC
+function static_tests() {
+  install_if_missing core/go gofmt
+
+  ( cd /src || return 1
+    local gofiles
+    gofiles=$(find ./ -maxdepth 3 -name '*.go')
```
[the remark this answers is not in the document]
Comment (jonsmorrow) [reply .studiorc]: will fix this.

File: .studiorc (hunk 4 above)
Comment (jonsmorrow) [reply .studiorc]: The escapes your seeing there are for re-writting the prior line to pass/fail output not color

**Pull request closed without merging**